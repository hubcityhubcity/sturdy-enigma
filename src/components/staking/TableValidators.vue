<template>
  <div>
    <table class="data-table">
      <thead>
        <PanelSort
          :sort="sort"
          :properties="properties"
          :show-on-mobile="showOnMobile"
        />
      </thead>
      <tbody
        v-infinite-scroll="loadMore"
        infinite-scroll-distance="400"
        name="flip-list"
      >
        <LiValidator
          v-for="(validator, index) in showingValidators"
          :key="validator.operatorAddress"
          :index="index"
          :validator="validator"
          :delegation="getDelegation(validator)"
          :rewards="getRewards(validator)"
          :show-on-mobile="showOnMobile"
        />
      </tbody>
    </table>
  </div>
</template>

<script>
import { mapGetters } from "vuex"
import orderBy from "lodash.orderby"
import LiValidator from "staking/LiValidator"
import PanelSort from "staking/PanelSort"
import gql from "graphql-tag"

export default {
  name: `table-validators`,
  components: {
    LiValidator,
    PanelSort
  },
  props: {
    validators: {
      type: Array,
      required: true
    },
    delegations: {
      type: Array,
      default: () => []
    },
    showOnMobile: {
      type: String,
      default: () => "returns"
    }
  },
  data: () => ({
    rewards: [],
    sort: {
      property: `expectedReturns`,
      order: `desc`
    },
    showing: 15
  }),
  computed: {
    ...mapGetters([`address`, `network`]),
    sortedEnrichedValidators() {
      return orderBy(
        this.validators.map(validator => ({
          ...validator,
          smallName: validator.name ? validator.name.toLowerCase() : ""
        })),
        [this.sort.property],
        [this.sort.order]
      )
    },
    showingValidators() {
      return this.sortedEnrichedValidators.slice(0, this.showing)
    },
    properties() {
      return [
        {
          title: `Status`,
          value: `status`,
          tooltip: `Validation status of the validator`
        },
        {
          title: `Name`,
          value: `smallName`,
          tooltip: `The validator's name`
        },
        {
          title: `Rewards`,
          value: `expectedReturns`,
          tooltip: `Approximate annualized reward`
        },
        {
          title: `Voting Power`,
          value: `votingPower`,
          tooltip: `Percentage of voting shares`
        }
      ]
    }
  },
  watch: {
    "sort.property": function() {
      this.showing = 15
    },
    "sort.order": function() {
      this.showing = 15
    }
  },
  mounted() {
    this.$apollo.queries.rewards.startPolling(1000 * 60 * 5)
  },
  methods: {
    loadMore() {
      this.showing += 10
    },
    getDelegation({ operatorAddress }) {
      return this.delegations.find(
        ({ validator }) => validator.operatorAddress === operatorAddress
      )
    },
    getRewards({ operatorAddress }) {
      return this.rewards.find(
        ({ validator }) => validator.operatorAddress === operatorAddress
      )
    }
  },
  apollo: {
    rewards: {
      query: gql`
        query Rewards($networkId: String!, $delegatorAddress: String!) {
          rewards(networkId: $networkId, delegatorAddress: $delegatorAddress) {
            validator {
              operatorAddress
            }
            amount
          }
        }
      `,
      skip() {
        return !this.address
      },
      variables() {
        return {
          networkId: this.network,
          delegatorAddress: this.address
        }
      },
      update: result => {
        return result.rewards || []
      }
    }
  }
}
</script>
<style scoped>
@media screen and (max-width: 550px) {
  .data-table td {
    overflow: hidden;
  }

  .data-table__row__info {
    max-width: 22rem;
  }
}

.flip-list-move {
  transition: transform 0.3s;
}

.data-table >>> th:first-child {
  width: 5%;
  color: var(--dim);
  font-size: var(--sm);
}

.data-table >>> th:nth-child(2) {
  width: 10%;
  color: var(--dim);
  font-size: var(--sm);
}

.data-table >>> th:nth-child(3) {
  width: 50%;
}
</style>
