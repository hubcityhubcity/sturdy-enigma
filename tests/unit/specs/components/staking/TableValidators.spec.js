import { shallowMount } from "@vue/test-utils"
import TableValidators from "src/components/staking/TableValidators"
import validators from "../../store/json/validators.js"

describe(`TableValidators`, () => {
  let wrapper, $store

  const getters = {
    committedDelegations: {
      [validators[0].operator_address]: 10
    },

    bondDenom: `stake`,

    lastHeader: {
      chain_id: `gaia-20k`,
      height: `6001`
    }
  }

  beforeEach(() => {
    $store = {
      commit: jest.fn(),
      dispatch: jest.fn(),
      state: {
        minting: {
          annualProvision: "100"
        },
        distribution: {
          loaded: true,
          rewards: {
            cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctqzh8yqw: {
              stake: 1000
            }
          }
        },
        pool: {
          pool: {
            bonded_tokens: 1000
          }
        },
        delegates: {
          signingInfos: {
            cosmosvaladdr15ky9du8a2wlstz6fpx3p4mqpjyrm5ctqzh8yqw: {
              missed_blocks_counter: 2
            }
          }
        },
        session: {
          address: `address1234`,
          signedIn: true
        }
      },
      getters: JSON.parse(JSON.stringify(getters)) // clone so we don't overwrite by accident
    }

    wrapper = shallowMount(TableValidators, {
      mocks: {
        $store
      },
      propsData: { validators },
      directives: {
        infiniteScroll: () => {}
      }
    })
    wrapper.setData({ rollingWindow: 10000 })
  })

  it(`shows a validator table`, async () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it(`should create an enriched validator object for a signed in user`, () => {
    expect(wrapper.vm.enrichedValidators[0].small_moniker).toBe(`mr_mounty`)
    expect(wrapper.vm.enrichedValidators[0].my_delegations).toBe(10)
    expect(wrapper.vm.enrichedValidators[0].rewards).toBe(1000)
    expect(wrapper.vm.enrichedValidators[0].expectedReturns).toBe(0.1)
  })

  it(`should create an enriched validator object for a user who is not signed in `, () => {
    wrapper.vm.session.signedIn = false
    expect(wrapper.vm.enrichedValidators[1].my_delegations).toBe(0)
    expect(wrapper.vm.enrichedValidators[1].rewards).toBe(0)
  })

  it(`should sort the delegates by selected property`, () => {
    wrapper.vm.sort.property = `operator_address`
    wrapper.vm.sort.order = `desc`

    expect(
      wrapper.vm.sortedEnrichedValidators.map(x => x.operator_address)
    ).toEqual(validators.map(x => x.operator_address))

    wrapper.vm.sort.property = `operator_address`
    wrapper.vm.sort.order = `asc`

    expect(
      wrapper.vm.sortedEnrichedValidators.map(x => x.operator_address)
    ).toEqual(validators.map(x => x.operator_address).reverse())
  })

  it(`should load more validators (infinite scrolling)`, async () => {
    wrapper.setData({ showing: 2 })
    expect(wrapper.findAll("livalidator-stub").length).toBe(2)
    wrapper.vm.loadMore()
    expect(wrapper.findAll("livalidator-stub").length).toBe(5)
  })

  it(`should update rewards on new blocks`, () => {
    const $store = { dispatch: jest.fn() }
    const newHeader = { height: `30` }
    TableValidators.watch.lastHeader.handler.call({ $store }, newHeader)
    expect($store.dispatch).toHaveBeenCalledWith(`getRewardsFromMyValidators`)
  })
})
