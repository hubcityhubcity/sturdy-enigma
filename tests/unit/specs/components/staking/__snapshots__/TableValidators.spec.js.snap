// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableValidators shows a validator table 1`] = `
<div>
  <table
    class="data-table"
  >
    <thead>
      <panelsort-stub
        properties="[object Object],[object Object],[object Object],[object Object]"
        showonmobile="returns"
        sort="[object Object]"
      />
    </thead>
     
    <tbody
      infinite-scroll-distance="400"
      name="flip-list"
    >
      <livalidator-stub
        delegation="[object Object]"
        index="0"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="1"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="2"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="3"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="4"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="5"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="6"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="7"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="8"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="9"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="10"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="11"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="12"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="13"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
      <livalidator-stub
        delegation="[object Object]"
        index="14"
        rewards=""
        showonmobile="returns"
        stakingdenom="ATOM"
        validator="[object Object]"
      />
    </tbody>
  </table>
</div>
`;
